modules/gapi/src/streaming/onevpl/accelerators/dx11_alloc_resource.cpp

```diff
@@ -0,0 +1,392 @@
+    return MFX_ERR_NONE;
+}
+
+mfxStatus DX11AllocationItem::exclusive_access_acquire_unsafe(mfxFrameData *ptr) {
+    GAPI_LOG_DEBUG(nullptr, "acquire WRITE lock: " << this);
+    GAPI_LOG_DEBUG(nullptr, "texture: " << get_texture() <<
+                            ", sub id: " << get_subresource());
+    D3D11_MAP mapType = D3D11_MAP_WRITE;
+    UINT mapFlags = D3D11_MAP_FLAG_DO_NOT_WAIT;
+
+    HRESULT err = S_OK;
+    D3D11_MAPPED_SUBRESOURCE lockedRect {};
+    do {
+        err = get_device_ctx()->Map(get_staging_texture(), 0, mapType, mapFlags, &lockedRect);
+        if (S_OK != err && DXGI_ERROR_WAS_STILL_DRAWING != err) {
+            GAPI_LOG_WARNING(nullptr, "Cannot Map staging texture in device context, error: " << std::to_string(HRESULT_CODE(err)));
+            return MFX_ERR_LOCK_MEMORY;
+        }
+    } while (DXGI_ERROR_WAS_STILL_DRAWING == err);
+
+    if (FAILED(err)) {
+        GAPI_LOG_WARNING(nullptr, "Cannot lock frame");
+        return MFX_ERR_LOCK_MEMORY;
+    }
+
+    D3D11_TEXTURE2D_DESC desc {};
+    get_texture()->GetDesc(&desc);
+    switch (desc.Format) {
+        case DXGI_FORMAT_NV12:
+            ptr->Pitch = (mfxU16)lockedRect.RowPitch;
+            ptr->Y = (mfxU8 *)lockedRect.pData;
+            ptr->UV = (mfxU8 *)lockedRect.pData + desc.Height * lockedRect.RowPitch;
+            if (!ptr->Y || !ptr->UV) {
+                GAPI_LOG_WARNING(nullptr, "DXGI_FORMAT_NV12 locked frame data is nullptr");
+                return MFX_ERR_LOCK_MEMORY;
+            }
+            break;
+        default:
+            GAPI_LOG_WARNING(nullptr, "Unknown DXGI format: " << desc.Format);
+            return MFX_ERR_LOCK_MEMORY;
+    }
+
+    GAPI_LOG_DEBUG(nullptr, "WRITE access granted: " << this);
+    return MFX_ERR_NONE;
+}
+
+mfxStatus DX11AllocationItem::exclusive_access_release_unsafe(mfxFrameData *ptr) {
+    GAPI_LOG_DEBUG(nullptr, "releasing WRITE lock: " << this);
+    GAPI_LOG_DEBUG(nullptr, "texture: " << get_texture() <<
+                            ", sub id: " << get_subresource());
+
+    get_device_ctx()->Unmap(get_staging_texture(), 0);
+
+    get_device_ctx()->CopySubresourceRegion(get_texture(),
+                                            get_subresource(),
+                                            0, 0, 0,
+                                            get_staging_texture(), 0,
+                                            nullptr);
+
+    if (ptr) {
+        ptr->Pitch = 0;
+        ptr->U = ptr->V = ptr->Y = 0;
+        ptr->A = ptr->R = ptr->G = ptr->B = 0;
+    }
+    GAPI_LOG_DEBUG(nullptr, "access on WRITE released: " << this);
+    return MFX_ERR_NONE;
+}
+
+DX11AllocationRecord::DX11AllocationRecord() = default;
+
+DX11AllocationRecord::~DX11AllocationRecord() {
+    GAPI_LOG_DEBUG(nullptr, "record: " << this <<
+                            ", subresources count: " << resources.size());
+
+    for (AllocationId id : resources) {
```

Comment:
Is this `AllocationId` already reference?

Reply:
it is pointer(please find declaration below)